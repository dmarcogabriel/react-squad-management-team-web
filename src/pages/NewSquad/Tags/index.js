import React, { useState, createRef, useEffect } from 'react';
import { MdClose } from 'react-icons/md';

const Tags = ({ defaultTags, onChange = () => {}, className }) => {
  const [tags, setTags] = useState([]);
  const [text, setText] = useState('');

  const textInputRef = createRef();

  const handleRemoveTag = (e) => {
    const newTags = tags.filter((t) => t !== e);
    setTags(newTags);
    onChange(newTags);
    setText('');
  };

  useEffect(() => {
    if (text.includes(';') || text.includes('\n')) {
      const [tag] = text.split(';');
      const newTags = [...tags, tag];

      setTags(newTags);
      onChange(newTags);
      setText('');
    }
  }, [text]);

  useEffect(() => {
    if (defaultTags && defaultTags.length) setTags(defaultTags);
  }, [defaultTags]);

  return (
    <div className={className}>
      <label htmlFor="tags">
        <p className="font-bold text-base">Tags</p>

        <button
          data-testid="button-field"
          type="button"
          onClick={() => textInputRef.current.focus()}
          className="flex rounded border-2 border-gray-400 my-3 py-2 w-full min-height-3 z-10"
        >
          <div data-testid="tags" className="flex flex-wrap relative">
            {tags.map((tag) => (
              <button
                data-testid={`tag-button-${tag}`}
                key={tag}
                type="button"
                className="flex justify-between items-center py-1 px-2 m-1 bg-primary-dark text-white text-sm rounded-full"
                onClick={() => handleRemoveTag(tag)}
              >
                <p className="inline mr-2">{tag}</p>

                <MdClose />
              </button>
            ))}

            <p data-testid="value" className="inline text-base">
              {text}
            </p>
          </div>

          <textarea
            data-testid="textarea"
            id="tags"
            ref={textInputRef}
            className="opacity-0 resize-none absolute z-0"
            value={text}
            onChange={({ target }) => setText(target.value)}
          />
        </button>
      </label>
    </div>
  );
};

export default Tags;
