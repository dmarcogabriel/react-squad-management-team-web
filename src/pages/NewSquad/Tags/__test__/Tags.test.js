import React from 'react';
import ReactDOM from 'react-dom';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Tags from '..';

const mockTags = ['test', 'test2', 'test3'];

describe('<Tags />', () => {
  it('render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Tags />, div);
  });

  it('execut field button click event', () => {
    const { getByTestId } = render(<Tags />);

    const button = getByTestId('button-field');
    userEvent.click(button);

    const textarea = getByTestId('textarea');

    expect(textarea).toHaveFocus();
  });

  it('execut tag button click event', () => {
    let result;
    const onChange = (e) => {
      result = e;
    };
    const expectedValue = mockTags.filter((tag) => tag !== 'test2');

    const { getByTestId } = render(
      <Tags defaultTags={mockTags} onChange={onChange} />
    );

    const button = getByTestId('tag-button-test2');
    userEvent.click(button);

    expect(result).toEqual(expectedValue);
  });

  it('execute textarea change event', () => {
    const { getByTestId } = render(<Tags />);

    const textarea = getByTestId('textarea');
    const value = getByTestId('value');

    fireEvent.change(textarea, { target: { value: 'Testing' } });

    expect(value).toHaveTextContent('Testing');
  });

  it('execute textarea create tag', () => {
    const mock = {
      onChange: (e) => e,
    };
    const spy = jest.spyOn(mock, 'onChange');

    const { getByTestId, rerender } = render(<Tags onChange={spy} />);

    const textarea = getByTestId('textarea');

    fireEvent.change(textarea, { target: { value: 'Testing;' } });

    rerender(<Tags onChange={spy} />);

    const createdTag = getByTestId('tags');

    expect(createdTag.childElementCount).toEqual(2);
    expect(spy).toHaveBeenCalled();
  });

  it('match snapshot', () => {
    expect(renderer.create(<Tags />).toJSON()).toMatchSnapshot();
  });
});
