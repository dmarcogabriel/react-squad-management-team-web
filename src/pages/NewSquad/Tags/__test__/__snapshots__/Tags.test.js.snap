// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tags /> match snapshot 1`] = `
<div>
  <label
    htmlFor="tags"
  >
    <p
      className="font-bold text-base"
    >
      Tags
    </p>
    <button
      className="flex rounded border-2 border-gray-400 my-3 py-2 w-full min-height-3 z-10"
      data-testid="button-field"
      onClick={[Function]}
      type="button"
    >
      <div
        className="flex flex-wrap relative"
        data-testid="tags"
      >
        <p
          className="inline text-base"
          data-testid="value"
        >
          
        </p>
      </div>
      <textarea
        className="opacity-0 resize-none absolute z-0"
        data-testid="textarea"
        id="tags"
        onChange={[Function]}
        value=""
      />
    </button>
  </label>
</div>
`;
